sirepo/auth/ldap.py

```diff
@@ -0,0 +1,87 @@
+"""LDAP authentication support
+:copyright: Copyright (c) 2018-2022 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern import pkconfig
+from pykern import pkinspect
+from pykern.pkdebug import pkdlog, pkdp
+from pykern.pkcollections import PKDict
+import ldap
+import re
+import sirepo.quest
+import sirepo.util
+
+# TODO(BG): Write test
+
+AUTH_METHOD = "ldap"
+
+AUTH_METHOD_VISIBLE = True
+
+_cfg = None
+
+_ESCAPE_DN_MAIL = re.compile(r"([,\\#<>;\"=+])")
+
+#: module handle
+this_module = pkinspect.this_module()
+
+#: Well known alias for auth
+user_model = "AuthEmailUser"
+
+
+class API(sirepo.quest.API):
+    @sirepo.quest.Spec("require_cookie_sentinel")
+    def api_authLdapLogin(self):
+
+        # validates username/password, escapes special chars for ldap dn
+        def _validate_escape_credentials(req):
```

Comment:
Functions are sorted alphabetically, put this after `_user`. The name is awkward. If you want something shorter, use `_validate_and_escape_creds`. The word `creds` in this context (nested function) is fine, because someone reading this will know what `creds` means here. For a public method, I would used credentials, but below I used `creds` for the object it returns.